### PR TITLE
WIP : AlphaBeta improvements

### PR DESCRIPTION
**Pull request by tyrcho**
#9

### CODE REVIEW

**Comment by huiwang** [comment]
@tyrcho did you obtain some metrics?

**Comment by tyrcho** [comment]
Yes, but this is not very relevant yet since I use a sample too stupid ... this needs a proper heuristic and moves sorting to be useful.

I got 23 ops/sec on depth 5 with an example of branching 20.

**Comment by huiwang** [comment]
@tyrcho is there anything i can help to finish this PR?

**Comment by huiwang** [comment]
@tyrcho i relaunched the build and it seems some tests are failing randomly.

File: src/main/scala/ai/scala/fp/game/alphabeta/AlphaBetaAi.scala [hunk 1]
```diff
@@ -2,41 +2,60 @@ package ai.scala.fp.game.alphabeta
...
-    val moves = rules.validMoves(state)
-    //TODO : can we cut branches also from root node ??
-    moves.minBy(m => negamax(rules.applyMove(state, m), depth - 1))
+    val sorted = sortedMoves(state)
+    best(sorted, MIN, MAX, state, depth).move.getOrElse(sorted.head)
+  }
+
+
+  def sortedMoves(state: S): Seq[M] = {
+    rules.validMoves(state).sortBy(m => heuristic(rules.applyMove(state, m)))
```
Comment (huiwang) [inline src/main/scala/ai/scala/fp/game/alphabeta/AlphaBetaAi.scala]: @tyrcho with sortBy, it's possible that the number of calls on  heuristic functions is greater than number of elements.  given that heuritic function can be heavy in computation in some cases, it's better to compute the heuristic first and then sort them.  refer to [MuPlusLambda](https://github.com/TrueLaurel/CodinGame-Scala-Kit/blob/master/src/main/scala/com/truelaurel/codingame/metaheuristic/evolutionstrategy/MuPlusLambda.scala#L24) or [GeneticAlgorithm](https://github.com/TrueLaurel/CodinGame-Scala-Kit/blob/master/src/main/scala/com/truelaurel/codingame/metaheuristic/genetic/GeneticAlgorithm.scala#L16)

Comment (tyrcho) [inline src/main/scala/ai/scala/fp/game/alphabeta/AlphaBetaAi.scala]: @huiwang I won't have time to finish this today and I'm not sure I can work on it during the WE. So I'm fine with you taking over from here and implementing the iterative deepening. I also agree with the sortBy optimization, it should be pretty straightforward to implement with a Map[M, Double].

Comment (huiwang) [inline src/main/scala/ai/scala/fp/game/alphabeta/AlphaBetaAi.scala]: Ok, I'll work on the iterative deepening part this weekend.  I had tried with the Map[M, Double] Solution. You know what? I spent too much time in hashing the M.

**Comment by huiwang** [comment]
@tyrcho I merged your improvements to avoid leaving it too long in a dev branch. I'll create a PR dedicated for iterative deepening.

File: src/main/scala/ai/scala/fp/game/alphabeta/AlphaBetaAi.scala (hunk 1 above)
Comment (tyrcho) [inline src/main/scala/ai/scala/fp/game/alphabeta/AlphaBetaAi.scala]: @huiwang I thought about it, would a `moves.map(m => (m,heuristic(m)).sortBy(_._2).map(_._1)` be faster ?

**Pull request merged**